contracts/Relay.sol

```diff
@@ -0,0 +1,431 @@
+            _anchor.hash256(),
+            _currentDigest);
+        return true;
+    }
+
+    /// @notice             Adds headers to storage after validating
+    /// @dev                We check integrity and consistency of the header chain
+    /// @param  _anchor     The header immediately preceeding the new chain
+    /// @param  _headers    A tightly-packed list of 80-byte Bitcoin headers
+    /// @return             True if successfully written, error otherwise
+    function addHeaders(bytes calldata _anchor, bytes calldata _headers) external returns (bool) {
+        return _addHeaders(_anchor, _headers, false);
+    }
+
+    /// @notice                 Adds headers to storage, performs additional validation of retarget
+    /// @dev                    Checks the retarget, the heights, and the linkage
+    /// @param  _oldPeriodStart The first header in the difficulty period being closed
+    /// @param  _oldPeriodEnd   The last header in the difficulty period being closed
+    /// @param  _headers        A tightly-packed list of 80-byte Bitcoin headers
+    /// @return                 True if successfully written, error otherwise
+    function addHeadersWithRetarget(
+        bytes calldata _oldPeriodStart,
+        bytes calldata _oldPeriodEnd,
+        bytes calldata _headers
+    ) external returns (bool) {
+        bytes memory _newPeriodStart = _headers.slice(0, 80);
+
+        /* NB: requires that both blocks are known */
+        uint256 _startHeight = _findHeight(_oldPeriodStart.hash256());
+        uint256 _endHeight = _findHeight(_oldPeriodEnd.hash256());
+
+        /* NB: retargets should happen at 2016 block intervals */
+        require(
+            _endHeight % 2016 == 2015,
+            "Must provide the last header of the closing difficulty period");
+        require(
+            _endHeight == _startHeight.add(2015),
+            "Must provide exactly 1 difficulty period");
+
+        /* NB: This comparison looks weird because header nBits encoding truncates targes */
+        uint256 _actualTarget = _newPeriodStart.extractTarget();
+        uint256 _expectedTarget = BTCUtils.retargetAlgorithm(
+            _oldPeriodStart.extractTarget(),
+            _oldPeriodStart.extractTimestamp(),
+            _oldPeriodEnd.extractTimestamp()
+        );
+        require(
+            (_actualTarget & _expectedTarget) == _actualTarget,
+            "Invalid retarget provided");
+
+        // Pass all but the first through to be added
+        return _addHeaders(_oldPeriodEnd, _headers, true);
+    }
+
+    /// @notice         Finds the height of a header by its digest
+    /// @dev            Will fail if the header is unknown
+    /// @param _digest  The header digest to search for
+    /// @return         The height of the header
+    function _findHeight(bytes32 _digest) internal view returns (uint256) {
+        uint256 _height = 0;
+        bytes32 _current = _digest;
+        for (uint256 i = 0; i < 5; i = i.add(1)) {
```

Comment:
Should this only search for 5 blocks after the current one?

Reply:
height is stored every 4, so if we traverse 5 and find nothing, it doesn't exist.

Reply:
OK so make `4` a constant at the top of the file and make the loop end at `constant+1`, the 5 looks arbitrary to a reader otherwise